apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,272 @@
+const Interpreter = require('@code-dot-org/js-interpreter');
+const codegen = require('../../../codegen');
+
+module.exports = class CustomMarshalingInterpreter extends Interpreter {
+
+  constructor(code, customMarshaler, opt_initFunc) {
+    super(code);
+    if (!customMarshaler) {
+      throw new Error("You must provide a CustomMarshaler to CustomMarshalingInterpreter");
+    }
+    this.customMarshaler = customMarshaler;
+
+    // TODO: because of the way the interpreter is written, we have to re-do
+    // this work from the parent class's constructor because it needs
+    // customMarshaler to have been set before-hand. Unfortunately, javascript
+    // does not allow us to set customMarshaler until after the parent constructor
+    // has finished.
+    this.initFunc_ = opt_initFunc;
+    this.globalScope = this.createScope(this.ast, null);
+    this.stateStack = [{
+      node: this.ast,
+      scope: this.globalScope,
+      thisExpression: this.globalScope,
+      done: false
+    }];
+  }
+
+  /**
+   * Helper to determine if we should prevent custom marshalling from occurring
+   * in a situation where we normally would use it. Allows us to block from a
+   * specific list of properties and a hardcoded list of instance types that are
+   * not safe to return into the interpreter sandbox.
+   *
+   * @param {string} name Name of property.
+   * @param {!Object} obj Data object.
+   * @param {Object} nativeParent Native parent object (if parented).
+   * @return {boolean} true if property access should be blocked.
+   */
+  shouldBlockCustomMarshalling_(name, obj, nativeParent) {
+    if (-1 !== this.customMarshaler.blockedProperties.indexOf(name)) {
+      return true;
+    }
+    var value = obj.isCustomMarshal ? obj.data[name] : nativeParent[name];
+    if (value instanceof Node || value instanceof Window) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Wrapper to Interpreter's getProperty (extended for custom marshaling)
+   *
+   * Fetch a property value from a data object.
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @return {!Object} Property value (may be UNDEFINED).
+   * @override
+   */
+  getProperty(obj, name) {
+    name = name.toString();
+    var nativeParent;
+    var customMarshalValue;
+    if (obj.isCustomMarshal) {
+      if (this.shouldBlockCustomMarshalling_(name, obj)) {
+        return this.UNDEFINED;
+      } else {
+        customMarshalValue = obj.data[name];
+      }
+    } else {
+      var hasProperty = false;
+      if (!obj.isPrimitive) {
+        hasProperty = super.hasProperty(obj, name);
+      }
+      if (!hasProperty &&
+          obj === this.globalScope &&
+          !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+          !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)) {
+        customMarshalValue = nativeParent[name];
```

Comment:
:laughing: I love that this one case of this method is not covered, especially because I can't really tell what this case _is_.  Something about getting a custom-marshaled property from the interpreter's global scope?